pkg/capabilities/consensus/ocr3/reduce/reduce_aggregator.go

```diff
@@ -0,0 +1,392 @@
+			}
+
+			(*currentState)[field.InputKey] = singleValue
+			report[field.OutputKey] = singleValue
+
+			if oldValue == nil || currDeviation > field.Deviation.InexactFloat64() {
+				shouldReport = true
+			}
+		}
+	}
+
+	// Add static additonal data to the report
+	for key, value := range a.config.AdditionalData {
+		report[key] = value
+	}
+
+	// TODO: name below better ----
+	vm, err := values.WrapMap(currentState)
+	if err != nil {
+		return nil, err
+	}
+	marshalledState := values.ProtoMap(vm)
+	metadata, err := proto.Marshal(marshalledState)
+	if err != nil {
+		return nil, err
+	}
+
+	var toWrap any
+	switch a.config.ReportFormat {
+	case "array":
+		toWrap = []map[string]any{report}
+	case "map":
+		toWrap = report
+	default:
+		// unsupported type
+		toWrap = map[string]any{}
+	}
+
+	wrappedReportsNeedingUpdates, err := values.NewMap(map[string]any{
+		a.config.OutputFieldName: toWrap,
+	})
+	if err != nil {
+		return nil, err
+	}
+	reportsProto := values.Proto(wrappedReportsNeedingUpdates)
+
+	// ----
+
+	a.lggr.Debugw("Aggregate complete", "shouldReport", shouldReport)
+	return &types.AggregationOutcome{
+		EncodableOutcome: reportsProto.GetMapValue(),
+		Metadata:         metadata,
+		ShouldReport:     shouldReport,
+	}, nil
+}
+
+func (a *reduceAggregator) initializeCurrentState(previousOutcome *types.AggregationOutcome) (*map[string]any, error) {
+	currentState := map[string]any{}
+	if previousOutcome != nil {
+		pb := &pb.Map{}
+		proto.Unmarshal(previousOutcome.Metadata, pb)
+		mv, err := values.FromMapValueProto(pb)
+		if err != nil {
+			return nil, err
+		}
+		mv.UnwrapTo(currentState)
+		if err != nil {
+			return nil, err
+		}
+	}
+	for _, field := range a.config.Fields {
+		if _, ok := currentState[field.InputKey]; !ok {
+			// TODO: set default type. user defined type?
+			currentState[field.InputKey] = nil
+			a.lggr.Debugw("initializing empty onchain state for feed", "fieldInputKey", field.InputKey)
+		}
+	}
+	a.lggr.Debugw("current state initialized", "state", currentState, "previousOutcome", previousOutcome)
+	return &currentState, nil
+}
+
+func (a *reduceAggregator) extractValues(observations map[ocrcommon.OracleID][]values.Value, aggregationKey string) []any {
+	events := map[string][]any{}
+	for nodeID, nodeObservations := range observations {
+		// TODO: check the following comment:
+		// we only expect a single observation per node
+		if len(nodeObservations) == 0 || nodeObservations[0] == nil {
+			a.lggr.Warnf("node %d contributed with empty observations", nodeID)
+			continue
+		}
+		if len(nodeObservations) > 1 {
+			a.lggr.Warnf("node %d contributed with more than one observation", nodeID)
+			continue
+		}
+		vmap, err := nodeObservations[0].Unwrap()
+		if err != nil {
+			a.lggr.Warnf("could not unwrap observation value from node %d: %v", nodeID, err)
+			continue
+		}
+
+		value := vmap.(map[string]any)[aggregationKey]
+
+		switch v := value.(type) {
+		case string:
```

Comment:
@justinkaseman I wonder if you'd be better off dealing with the type complexity as low down as possible, rather than unwrapping the types here and dealing with it again further down.

Concretely, you just pass the values.Value to either mode or median and let those functions figure out what types they support + how they aggregate.

Eg. for mode, you can just do what the identical aggregator does and marshal the proto to some bytes and sha it. For median, you can just convert each type to a decimal (I think that's what we agreed on anyway in our last chat) and return an error if any one of the values can't be handled.

Reply:
That does seem cleaner. I'll try that.